File: DS/slidingWindow/longestSubString.js
// 15. Longest Substring Without Repeating Characters

// Problem Statement:
// Given a string, determine the length of the longest substring that does not contain any repeated characters.

// Input:

// String = "abcabcbb"

// Output:

// 3

function longestSubString(str) {
    let maxLength = 0;
    let charSet = new Set();
    let left = 0;

    for (let right = 0; right < str.length; right++) {
        while (charSet.has(str[right])) {
            charSet.delete(str[left]);
            left = left + 1;

        }

        charSet.add(str[right]);

       // let currentWindowSize = right -left + 1;
       // max

               maxLength = Math.max(maxLength, right - left + 1)






    }

    

}

console.log(longestSubString('abcabcbb'));

File: DS/slidingWindow/LongestSubString.java
import java.util.HashSet;

public class LongestSubString {
    public static int longestSubString(String str) {

         int maxLength = 0;
    int charSet = new Set();
    int left = 0;

        HashSet<Character> charSet = new HashSet<>();

        for (int right = 0; right < str.length; right++) {
        while (charSet.contains(str[right])) {
            charSet.remove(str[left]);
            left = left + 1;

        }

        charSet.add(str[right]);
        maxLength = Math.max(maxLength, right - left + 1)

       // let currentWindowSize = right -left + 1;

    }

    public static void main(String[] args){

    }

}